File: Challenges/450_DSA_Cracker/Matrix/010_Spiral_Traversal.py
#User function Template for python3

class Solution:
    def spirallyTraverse(self,matrix, r, c): 
        layer = 0
        res = []
        
        layers = min(r, c)//2
        
        while layer < layers: 
            for i in range(layer, c-1-layer) :
                res.append(matrix[layer][i])
            
            for i in range(layer, r-1-layer):
                res.append(matrix[i][c-1-layer])
            
            for i in range(c-1-layer, layer, -1):
                res.append(matrix[r-1-layer][i])
            
            for i in range(r-1-layer, layer, -1):
                res.append(matrix[i][layer])
            layer += 1
        
        if r<c and r%2==1 : 
            for i in range(layer, c-layer):
                res.append(matrix[r//2][i])
        if c<r and c%2==1:
            for i in range(layer, r-layer):
                res.append(matrix[i][c//2])
        if c==r and c%2==1 and r%2==1 : 
            res.append(matrix[r//2][c//2])
        return res

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        r,c = map(int, input().strip().split())
        values = list(map(int, input().strip().split()))
        k = 0
        matrix =[]
        for i in range(r):
            row=[]
            for j in range(c):
                row.append(values[k])
                k+=1
            matrix.append(row)
        obj = Solution()
        ans = obj.spirallyTraverse(matrix,r,c)
        for i in ans:
            print(i,end=" ")
        print()

# } Driver Code Ends

File: Challenges/450_DSA_Cracker/Matrix/010_Spiral_Traversal.java

//{ Driver Code Starts
import java.io.*;
import java.util.*;
class GFG
{
    public static void main(String args[])throws IOException
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            int r = sc.nextInt();
            int c = sc.nextInt();
            
            int matrix[][] = new int[r][c];
            
            for(int i = 0; i < r; i++)
            {
                for(int j = 0; j < c; j++)
                 matrix[i][j] = sc.nextInt();
            }
            Solution ob = new Solution();
            ArrayList<Integer> ans = ob.spirallyTraverse(matrix, r, c);
            for (Integer val: ans) 
                System.out.print(val+" "); 
            System.out.println();
        }
    }
}
// } Driver Code Ends


class Solution
{
    //Function to return a list of integers denoting spiral traversal of matrix.
    static ArrayList<Integer> spirallyTraverse(int matrix[][], int r, int c)
    {
        ArrayList<Integer> res = new ArrayList<Integer>();
        int layers = Math.min(r, c)/2, layer = 0;
        
        while(layer<layers) {
            for(int i=layer; i<c-1-layer; i++) 
                res.add(new Integer(matrix[layer][i]));
            
            for(int i=layer; i<r-1-layer; i++) 
                res.add(new Integer(matrix[i][c-1-layer]));
            
            for(int i=c-1-layer; i>layer; i--)
                res.add(new Integer(matrix[r-1-layer][i]));
            
            for(int i=r-1-layer; i>layer; i--)
                res.add(new Integer(matrix[i][layer]));
            layer++;
        }
        
        if(r==c && r%2==1 && c%2==1)
            res.add(new Integer(matrix[r/2][c/2]));
        else if(r<c && r%2==1)
            for(int i=layer; i<c-layer; i++)
                res.add(new Integer(matrix[r/2][i]));
        else if(c<r && c%2==1)
            for(int i=layer; i<r-layer; i++)
                res.add(new Integer(matrix[i][c/2]));
        return res;
    }
}
